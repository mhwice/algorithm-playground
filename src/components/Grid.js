import React, { useState, useEffect } from 'react';
import GridCell from './GridCell';

const createMatrix = (rows, columns) => {
  if (typeof rows !== 'number' || typeof columns !== 'number' || rows < 1 || columns < 1) {
    return [[0]]
  }
  let matrix = [];
  for (let i = 0; i < rows; i++) {
    let row = [];
    for (let j = 0; j < columns; j++) {
      row.push(i * columns + j);
    }
    matrix.push(row);
  }
  return matrix;
};

const Grid = ({ rows = 1, columns = 1}) => {

  const [matrix, setMatrix] = useState([[]]);

  useEffect(() => {
    setMatrix(createMatrix(rows, columns));
  }, []);

  return (
    <div className="grid">
      {
        matrix.map((row, rowIndex) => {
          return (
            <div className="row" key={`row${rowIndex}`}>
              {
                row.map((col) => <GridCell key={col.toString()} />)
              }
            </div>
          )
        })
      }
    </div>
  );
};

export { createMatrix, Grid as default };