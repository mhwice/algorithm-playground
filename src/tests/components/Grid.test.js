import React from 'react';
import { shallow } from 'enzyme';
import Grid, { createMatrix } from './../../components/Grid';

test('should render Grid correctly', () => {
  const wrapper = shallow(<Grid />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Grid correctly with props', () => {
  const wrapper = shallow(<Grid rows={3} columns={7} />);
  expect(wrapper).toMatchSnapshot();
});

test('should generate matrix correctly with valid input', () => {
  const matrix = createMatrix(2, 3);
  expect(matrix).toEqual([
    [0,1,2],
    [3,4,5]
  ]);
});

test('should generate matrix correctly with no input', () => {
  const matrix = createMatrix();
  expect(matrix).toEqual([
    [0]
  ]);
});

test('should generate matrix correctly with invalid input', () => {
  const matrix = createMatrix('apples', 3);
  expect(matrix).toEqual([
    [0]
  ]);
});