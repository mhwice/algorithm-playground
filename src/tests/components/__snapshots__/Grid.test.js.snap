// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Grid correctly 1`] = `
<div
  className="grid"
>
  <div
    className="row"
    key="row0"
  />
</div>
`;

exports[`should render Grid correctly with props 1`] = `
<div
  className="grid"
>
  <div
    className="row"
    key="row0"
  />
</div>
`;
